pkg/tcpip/transport/tcp/connect.go

```diff
@@ -927,7 +935,10 @@ func (e *endpoint) protocolMainLoop(handshake bool) *tcpip.Error {
 
 	// Tell waiters that the endpoint is connected and writable.
 	e.mu.Lock()
-	e.state = StateEstablished
+	if e.state != StateEstablished {
+		e.stack.Stats().TCP.CurrentEstablished.Increment()
+		e.state = StateEstablished
+	}
 	drained := e.drainDone != nil
```

Comment:
Check redundant.

Reply:
No problem to remove the check in accept.go. But for connect.go, if we remove the check, the test cannot pass:

test/syscalls/linux/proc_net.cc:186: Failure
Expected equality of these values:
  oldCurrEstab
    Which is: 0
  newCurrEstab - 2
    Which is: 1
test/syscalls/linux/proc_net.cc:217: Failure
Expected equality of these values:
  oldCurrEstab
    Which is: 0
  newCurrEstab
    Which is: 1

Reply:
Hmm that's odd. I will take a look later.